### PR TITLE
Site Editor: Prevent unintended actions on the classic theme

### PR DESCRIPTION
**Pull request by t-hamano**
Related to #54066, #52150
Related core ticket: https://core.trac.wordpress.org/ticket/58827
Related core PR: https://github.com/WordPress/wordpress-develop/pull/5201#issuecomment-1726991594

## What?
This PR prevents the following three unintended actions in case the Classic theme is allowed to access the Patterns page of the Site Editor.

- Can create a template part
- Can back to Design page
- Can access All template parts page
- The back button link on the All Template Parts page is to the dashboard even if the Site Editor's Patterns page is accessed directly (in this scenario, it is expected to return to the Patterns page)

| Before | After |
|--------|--------|
| before | after |

## Why?

As a first step toward making the Patterns page (`site-editor.php?path=/patterns`) accessible to all classic themes in the future, we are considering the following in WP6.4:

1. Add an `Appearance -> Patterns` link for all classic themes that point to `edit.php?post_type=wp_block`.
2. Allow `site-editor.php?path=/patterns` access for classic themes internally.

This means that users of the classic theme will not see the link to the Pattern page (`site-editor.php?path=/patterns`) anywhere, but will be able to access it by typing its URL directly into their browser.

Therefore, actions that are not expected in that case need to be prevented in advance.

## Testing Instructions

To temporarily allow access to the site editor from the classic theme, comment out [here](https://github.com/WordPress/wordpress-develop/blob/54a177910a80083930ee1284f648c1db5bf2e5c0/src/wp-admin/site-editor.php#L22-L32) in the WordPress core

1. Access `http://localhost:8889/wp-admin`
2. Activate Twenty Twenty One
3. Enter `http://localhost:8889/wp-admin/site-editor.php?path=%2Fpatterns` in the URL bar of your browser to access the Patterns page
4. Make sure the "Manage all template parts" menu is not displayed
5. Confirm that new template parts cannot be created
6. Check these steps also for the classic theme (EmptyHybrid) with block-template-parts support

### Additional Test

If a classic theme that supports block template parts goes directly to the Patterns page in the Site Editor, the back button link on the All Template Parts page should always go to the Patterns page instead of the Dashboard.

To control this, I introduce a URL parameter called `didAccessPatternsPage`.

Based on this, please test the following:

- Block Theme: It should work as before. No matter what you do, the URL should always have no `didAccessPatternsPage` parameter.
- Classic Theme _with_ template parts support:
  - Activate Emptyhybrid theme.
  - Access `http://localhost:8889/wp-admin/site-editor.php?path=%2Fpatterns` directly.
  - No matter what operation is performed, the All Template Parts page should return to the Patterns page.
  - The `didAccessPatternsPage` URL parameter is retained even if the browser is reloaded.
  - You should always return to the Patterns page, even after reloading the browser or reviving a closed tab with `Cmd+Shift+T`.
- Classic Theme _without_ template parts support:
  - Activate Twenty Twenty-One.
  - Access `http://localhost:8889/wp-admin/site-editor.php?path=%2Fpatterns` directly.
  - The URL should not have a `didAccessPatternsPage` parameter, no matter what operation is performed, not to mention that the template parts are not accessible.

### CODE REVIEW

**Review by kevin940726 — APPROVED** [review]
Makes sense to me! Thanks!

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for continuing the work on this one @t-hamano 👍 

The correct links and menu items are showing for me depending on the type of theme.

| Classic | Hybrid |
|---|---|
| Screenshot 2023-09-14 at 6 14 11 pm | Screenshot 2023-09-14 at 6 13 56 pm |

The only odd thing I noticed was when using a hybrid theme, you can access the template part index page via two ways. 

1. Appearance > Template Parts
2. Patterns > Manage all template parts

Regardless of which way you navigate there, hitting back will take you to wp-admin.

Is there a way we can take the user back to the Patterns page if that is how they got to the template parts index page in the site-editor?

**Comment by t-hamano** [comment]
> The only odd thing I noticed was when using a hybrid theme, you can access the template part index page via two ways. 
> 
> 1. Appearance > Template Parts
> 2. Patterns > Manage all template parts
> 
> Regardless of which way you navigate there, hitting back will take you to wp-admin.

My understanding is that this is expected behavior at this point.

As discussed in #54066, all Patterns links are `edit.php?post_type=wp_block` instead of `site-editor.php?path=/patterns` until the minimum WP version supported by the Gutenberg plugin increases to 6.4. Until then, we are only allowing access to the Patterns page (`site-editor.php?path=/patterns`) internally, so unless the user directly types `site-editor.php?path=/patterns` into the browser, the Patterns page will not be accessed.

In other words, there is only one way to access the template parts index page: "Appearance > Template Parts". Also, I don't think it is expected at this point to return to the patterns page from the template parts index page.

**Comment by aaronrobertshaw** [comment]
> Until then, we are only allowing access to the Patterns page (site-editor.php?path=/patterns) internally, so unless the user directly types site-editor.php?path=/patterns into the browser, the Patterns page will not be accessed.

If it can be accessed, even if only being typed directly, I'd expect users will. After all, that's how some users have been accessing the `edit.php?post_type=wp_block` page to manage patterns.

We could say there is only one official way to access the template parts index page, but there are two possible ways. So if there is a way we could return to the correct location using the history state or something, that would be a proactive step to avoiding a jarring experience.

**Comment by t-hamano** [comment]
> We could say there is only one official way to access the template parts index page, but there are two possible ways. So if there is a way we could return to the correct location using the history state or something, that would be a proactive step to avoiding a jarring experience.

I see, it checks if the page just before accessing the template parts index page was a Pattern page (i.e. whether the user accessed directly to the Patterns page) and if so, back to the Pattern page instead of the dashboard. This is the ideal behaviour, right?

**Comment by t-hamano** [comment]
I was wondering what approach would be best to determine if the Patterns page was accessed directly. At first, I considered adding that information to the URL parameter and making a determination based on that parameter, but the URL parameter would be overwritten in various places.

In the end, I took the following approach:

- Added a private API to determine whether the Patterns page has been accessed
- Dispatch the state that the Pattern page was accessed via `setDidAccessPatternsPage` when the Patterns page is rendered
- On the "All Template Parts" page, determine whether the Patterns page has already been accessed via the `didAccessPatternsPage` selector (that is, whether the Patterns page was accessed directly). Controls whether to return to the Patterns page or the dashboard based on the results.

With this logic, a classic theme with block-template-parts should behave like this:

- If you access the Patterns page directly, the "All template Parts" page will always return you to the Patterns page, no matter what page transitions you make, unless you reload your browser.
- When you go to Appearance > Template Parts, the "All Template Parts page" always returns you to the dashboard.

<details><summary>Screencast</summary>

https://github.com/WordPress/gutenberg/assets/54422211/0f4f5ce2-4c18-4b2c-bfdf-bd3bffcddeb8

</details>

**Comment by aaronrobertshaw** [comment]
> At first, I considered adding that information to the URL parameter and making a determination based on that parameter, but the URL parameter would be overwritten in various places.

I was expecting we'd do something along these lines so that reloading the page didn't interfere with going back to the correct location. From memory there is some two way syncing of the URL that can make things tricky. It should be doable though as we maintain some query string params for navigating pattern pages.

> Added a private API to determine whether the Patterns page has been accessed

This is an improvement but I do wonder if it is adding too much complexity for a pretty contrived edge case. It might be good to get some fresh eyes here. Perhaps @glendaviesnz and @kevin940726 might have some thoughts?

Thanks for your patience here again @t-hamano 👍

**Comment by kevin940726** [comment]
There's a more common way of doing this in general via [`history.state`](https://developer.mozilla.org/en-US/docs/Web/API/History/state). It's similar to how "modal routes" are implemented in some popular websites (Facebook, Instagram, Reddit, etc).

The trick is to append a `state` object with the current (`/patterns`) location when navigating after clicking the "Manage all template parts" link. Then, when clicking the "Back" button, we first check if `history.state` has the `/patterns` location, and just call `history.back()` if it has. On the other hand, if it's coming from a direct link (no `history.state`), then we just navigate the user to `wp-admin`.

The benefit of storing the information in `history.state` as opposed to local store is that it survives after reloading and even after closing and reviving the tab (`Cmd+Shift+T`).

However, as pointed out by @aaronrobertshaw, the two-way syncing gotcha in the site editor might make this more complicated than it should be though. 🤔

**Comment by t-hamano** [comment]
> The trick is to append a `state` object with the current (`/patterns`) location when navigating after clicking the "Manage all template parts" link. Then, when clicking the "Back" button, we first check if `history.state` has the `/patterns` location, and just call `history.back()` if it has. On the other hand, if it's coming from a direct link (no `history.state`), then we just navigate the user to `wp-admin`.

I think the concern with this approach is that the user has other ways to navigate away from the Patterns page other than clicking on "Manage all template parts" link.

Users might type "header" in the Command Palette to go to the template part's editing canvas. Or users might click "Create pattern" to go to the pattern editing canvas.

So I'm concerned about whether we can control all of these actions by using history.state. 🤔

**Comment by t-hamano** [comment]
> I was expecting we'd do something along these lines so that reloading the page didn't interfere with going back to the correct location. From memory there is some two way syncing of the URL that can make things tricky. It should be doable though as we maintain some query string params for navigating pattern pages.

I tried working with this approach. `history.state` alone probably didn't work in a bidirectional relationship with the URL, so I'm explicitly giving it `didAccessPatternsPage=1` as a URL parameter. This parameter is only added for classic themes that support block template parts.

It should work like this:

- Block Theme: It should work as before. No matter what you do, the URL should always have no `didAccessPatternsPage` parameter.
- Classic Theme _with_ template parts support:
  - Activate Emptyhybrid theme.
  - Access `http://localhost:8889/wp-admin/site-editor.php?path=%2Fpatterns` directly.
  - No matter what operation is performed, the All Template Parts page should return to the Patterns page.
  - The `didAccessPatternsPage` URL parameter is retained even if the browser is reloaded.
  - You should always return to the Patterns page, even after reloading the browser or reviving a closed tab with `Cmd+Shift+T`.
- Classic Theme _without_ template parts support:
  - Activate Twenty Twenty-One.
  - Access `http://localhost:8889/wp-admin/site-editor.php?path=%2Fpatterns` directly.
  - The URL should not have a `didAccessPatternsPage` parameter, no matter what operation is performed, not to mention that the template parts are not accessible.

<details><summary>Screencast</summary>

https://github.com/WordPress/gutenberg/assets/54422211/93e360fc-9c83-40df-a96b-c06c7de0e3d5

</details> 

I don't know if this approach is ideal, but I'd like your input.

File: packages/edit-site/src/components/sidebar-navigation-screen-patterns/index.js [hunk 1]
```diff
@@ -93,8 +96,21 @@ export default function SidebarNavigationScreenPatterns() {
...
+	);
+	const isTemplatePartsMode = useSelect( ( select ) => {
+		const settings = select( editSiteStore ).getSettings();
+		return !! settings.supportsTemplatePartsMode;
+	}, [] );
+	const templatePartsLink = useLink( {
+		path: '/wp_template_part/all',
+		// If a classic theme that supports template parts accessed
+		// the Patterns page directly, preserve that state in the URL.
+		didAccessPatternsPage: isTemplatePartsMode ? 1 : undefined,
```
Comment (kevin940726) [inline packages/edit-site/src/components/sidebar-navigation-screen-patterns/index.js]: Nit: I know `isTemplatePartsMode` implicitly implies `! isBlockBasedTheme` but to me it's not that obvious from the naming alone 😅 . Do you think we should add `! isBlockBasedTheme` just to make it clearer?

**Review by kevin940726 — APPROVED** [review]
Works well in my testing! I don't know how important `didAccessPatternsPage` is at this stage as there's not a public link, but I appreciate the thorough work around making sure the UX is as smooth as possible! Thank you! 💯

**Review by glendaviesnz — APPROVED** [review]
Tested as advertised for me as well:

With Classic theme:
✅  Could enter http://localhost:8889/wp-admin/site-editor.php?path=%2Fpatterns in the URL bar of  browser to access the Patterns page
✅ "Manage all template parts" menu is not displayed
✅ New template parts cannot be created

With Hybrid theme:
✅  Could enter http://localhost:8889/wp-admin/site-editor.php?path=%2Fpatterns in the URL bar of browser to access the Patterns page
✅ "Manage all template parts" menu was displayed and worked as expected

File: packages/edit-site/src/components/sidebar-navigation-screen-patterns/index.js (hunk 1 above)
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar-navigation-screen-patterns/index.js]: When someone comes to this fresh without much context, I think it could really help to be explicit here and probably doesn't hurt much to make this tweak.

File: packages/core-commands/src/site-editor-navigation-commands.js [hunk 2]
```diff
@@ -136,6 +136,11 @@ const getNavigationCommandLoaderPerPostType = ( postType ) =>
 const getNavigationCommandLoaderPerTemplate = ( templateType ) =>
 	function useNavigationCommandLoader( { search } ) {
 		const history = useHistory();
+		const {
+			params: { path, postType, didAccessPatternsPage },
+		} = useLocation();
```
Comment (aaronrobertshaw) [inline packages/core-commands/src/site-editor-navigation-commands.js]: This appears to throw an error when using the command palette anywhere except the site editor e.g. post editor, editing custom post type such as the wp_block patterns etc.

**Review by aaronrobertshaw — COMMENTED** [review]
As @kevin940726 and @glendaviesnz have noted, the navigation once within the site editor looks pretty solid. 

Unfortunately, an error gets thrown when trying to search commands in the command palette in the post editor.

Screenshot 2023-09-22 at 1 59 44 pm

If you have the time it might help to freshen up the PR description and test instructions along the lines of your [comment above](https://github.com/WordPress/gutenberg/pull/54422#issuecomment-1722133313) 🙏 

Special thanks for the video, I might have missed retesting the command palette without it jogging my memory 🙂

File: packages/core-commands/src/site-editor-navigation-commands.js (hunk 2 above)
Comment (aaronrobertshaw) [reply packages/core-commands/src/site-editor-navigation-commands.js]: We only need to check for these params while in the site editor, so a simple guard against the undefined location might suffice. What do you think?

Comment (kevin940726) [reply packages/core-commands/src/site-editor-navigation-commands.js]: Good spot! Guarding against `undefined` seems logical to me 👍

File: packages/edit-site/src/components/sidebar-navigation-screen-patterns/index.js (hunk 1 above)
Comment (t-hamano) [reply packages/edit-site/src/components/sidebar-navigation-screen-patterns/index.js]: I see, added by b3963ef 👍

File: packages/core-commands/src/site-editor-navigation-commands.js (hunk 2 above)
Comment (t-hamano) [reply packages/core-commands/src/site-editor-navigation-commands.js]: ~~Thank you for finding it! I fixed it with fa499af.~~

Fixed with fa499af and 58f0d0a.

**Comment by t-hamano** [comment]
Thank you everyone for your cooperation 🙇 I think this PR is ready to be merged. The PR description has also been updated.

However, [a PR for WP Core](https://github.com/WordPress/wordpress-develop/pull/5201) that allows access to the Site Editor from classic themes has not yet been committed.

Should this PR be merged after it is committed in the core? My understanding is that this is a bug fix that prevents unintended behavior, so it should be able to be backported after the Beta1 release.

Regarding the core PR, I'm also[ inviting the core team to review](https://wordpress.slack.com/archives/C055Y7FKS7N/p1695349591760369) on the `#6-4-release-leads` channel.

**Comment by kevin940726** [comment]
Tested well for me!

> Should this PR be merged after it is committed in the core? My understanding is that this is a bug fix that prevents unintended behavior, so it should be able to be backported after the Beta1 release.

I think it's okay to merge this before the core patch. It can be seen as a bug fix and doesn't necessarily depend on the core patch. As long as merging it alone doesn't break any existing behavior, I think we can just yolo merge this. Worst case is reverting it, not a big deal to me 🤷‍♂️.

> Regarding the core PR, I'm also[ inviting the core team to review](https://wordpress.slack.com/archives/C055Y7FKS7N/p1695349591760369) on the `#6-4-release-leads` channel.

Thank you! Much appreciated!

**Comment by t-hamano** [comment]
Update: The core ticket milestone was assigned 6.4. As @kevin940726 said, just merging this should have no effect on existing behavior. Also, even if the core patch is not applied in WP6.4, I would like to merge this PR as it should be useful in the future!

**Comment by mikachan** [comment]
I just cherry-picked this PR to the release/16.7 branch to get it included in the next release: d7cd5cd758
